### PR TITLE
test(core): add unit tests for `merge` trait

### PR DESCRIPTION
**Pull request by pluveto**
### Description

This update adds unit tests for the `Merge` trait implementations on `BTreeMap`, `BTreeSet`, `Vec`, and tuples. It also includes tests for the `is_empty` and `take` methods to ensure their correct functionality.

### Notes to the Reviewers

- The tests cover basic scenarios for merging collections and checking their emptiness.
- Consider the potential method name conflict with `BTreeSet`'s existing `take` method.

### Changelog Notice

- Added unit tests for merging functionality and utility methods (`is_empty`, `take`) on various collection types.

#### All Submissions:

* [x] I've signed all my commits.
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md).
* [x] I ran `cargo fmt` and `cargo clippy` before committing.

### CODE REVIEW

File: crates/core/tests/test_merge.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#[test]
+fn test_is_empty() {
+    let map: BTreeMap<i32, i32> = BTreeMap::new();
+    assert!(map.is_empty());
+
+    let set: BTreeSet<i32> = BTreeSet::new();
+    assert!(set.is_empty());
+
+    let vec: Vec<i32> = Vec::new();
+    assert!(vec.is_empty());
```
Comment (ValuedMammal) [inline crates/core/tests/test_merge.rs]: Should we qualify these with the trait method also?

```suggestion
    assert!(Merge::is_empty(&vec);
```

File: crates/core/tests/test_merge.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#[test]
+fn test_is_empty() {
+    let map: BTreeMap<i32, i32> = BTreeMap::new();
+    assert!(map.is_empty());
+
+    let set: BTreeSet<i32> = BTreeSet::new();
+    assert!(set.is_empty());
+
+    let vec: Vec<i32> = Vec::new();
+    assert!(Merge::is_empty(&vec);
```
Comment (LagginTimes) [inline crates/core/tests/test_merge.rs]: Missed the closing delimiter here:

```suggestion
    assert!(Merge::is_empty(&vec));
```

**Review by LagginTimes — COMMENTED** [review]
Would it be valuable to include tests for conflict scenarios as well? For example:

```
let mut map1: BTreeMap<i32, &str> = BTreeMap::new();
map1.insert(1, "a");
let mut map2: BTreeMap<i32, &str> = BTreeMap::new();
map2.insert(1, "b");

map1.merge(map2);

assert_eq!(map1.len(), 1);
assert_eq!(map1.get(&1), Some(&"b"));
```

File: crates/core/tests/test_merge.rs (hunk 2 above)
Comment (ValuedMammal) [reply crates/core/tests/test_merge.rs]: I would apply the same suggestion on L59 and L62 to disambiguate from `BTreeMap::is_empty` and `BTreeSet::is_empty`.

**Comment by pluveto** [comment]
I've added refactored empty check assertions and merge conflict tests. Squashed.

File: crates/core/tests/test_merge.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    set2.insert(1);
+
+    set1.merge(set2);
+
+    assert_eq!(set1.len(), 1);
+    assert!(set1.contains(&1));
+}
+
+#[test]
+fn test_vec_merge_conflict() {
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
fn test_vec_merge_duplicates() {
```

**Review by evanlinjin — APPROVED** [review]
ACK 871df7bb24b6c15a096d6b98bdfc7d6877cbd29c

File: crates/core/tests/test_merge.rs (hunk 3 above)
Comment (pluveto) [reply crates/core/tests/test_merge.rs]: Updated.

**Comment by LagginTimes** [comment]
ACK 665273dae475b3503c25aa92f77004f7cd16b20b

File: crates/core/tests/test_merge.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    let vec: Vec<i32> = Vec::new();
+    assert!(<Vec<i32>>::is_empty(&vec));
+}
+
+#[test]
+fn test_take() {
+    let mut map: BTreeMap<i32, i32> = BTreeMap::new();
+    map.insert(1, 1);
+    let taken_map = map.take();
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    let taken_map = Merge::take(&mut map);
```
This makes it more explicit where we are calling `take` from. Could you please do this for all calls to `Merge` methods?

File: crates/core/tests/test_merge.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    assert!(taken_map.is_some());
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    assert_eq!(taken_map, Some([(1, 1)].into()));
```
Instead of just checking whether something exists or does not exist, we should check the expected value. Please check all `assert!()` calls to see whether it makes more sense to be `assert_eq!()`.

File: crates/core/tests/test_merge.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+fn test_btree_set_merge_conflict() {
+    let mut set1: BTreeSet<i32> = BTreeSet::new();
+    set1.insert(1);
+    let mut set2: BTreeSet<i32> = BTreeSet::new();
+    set2.insert(1);
+
+    set1.merge(set2);
+
+    assert_eq!(set1.len(), 1);
+    assert!(set1.contains(&1));
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    assert_eq!(set1, [1].into());
```

**Review by evanlinjin — CHANGES_REQUESTED** [review]
Great test additions. Here are some suggestions to make the tests more complete/easier to read.

**Comment by pluveto** [comment]
Indeed, that's more readable. Updated.

File: crates/core/tests/test_merge.rs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    let expected_vec = vec![1, 2, 3, 4];
+    assert_eq!(tuple1.0, expected_vec);
+    let expected_set: BTreeSet<i32> = BTreeSet::from([3, 4]);
+    assert_eq!(tuple1.1, expected_set);
+}
+
+#[test]
+fn test_is_empty() {
+    let map: BTreeMap<i32, i32> = BTreeMap::new();
+    assert!(<BTreeMap<i32, i32>>::is_empty(&map));
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]: This is testing the `BTreeMap::is_empty` method which is part of the stdlib. We want to test the `Merge::is_empty` impls on these types.

```suggestion
    assert!(Merge::is_empty(&map));
```

File: crates/core/tests/test_merge.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    let set: BTreeSet<i32> = BTreeSet::new();
+    assert!(<BTreeSet<i32>>::is_empty(&set));
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    assert!(Merge::is_empty(&set));
```

File: crates/core/tests/test_merge.rs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    let vec: Vec<i32> = Vec::new();
+    assert!(<Vec<i32>>::is_empty(&vec));
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    assert!(Merge::is_empty(&vec));
```

File: crates/core/tests/test_merge.rs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+fn test_take() {
+    let mut map: BTreeMap<i32, i32> = BTreeMap::new();
+    map.insert(1, 1);
+    let taken_map = Merge::take(&mut map);
+    assert_eq!(taken_map, Some(BTreeMap::from([(1, 1)])));
+    assert!(map.is_empty());
+
+    let mut set: BTreeSet<i32> = BTreeSet::new();
+    set.insert(1);
+    let taken_set = <BTreeSet<i32> as Merge>::take(&mut set);
```
Comment (evanlinjin) [inline crates/core/tests/test_merge.rs]:
```suggestion
    let taken_set = Merge::take(&mut set);
```

**Review by evanlinjin — DISMISSED** [review]
Thanks for pushing forward on this. Let me know if you have any questions.

File: crates/core/tests/test_merge.rs (hunk 10 above)
Comment (pluveto) [reply crates/core/tests/test_merge.rs]: Thanks, fixed.

**Comment by pluveto** [comment]
Fixed and Rebased

**Review by oleonardolima — COMMENTED** [review]
cACK a53f07be1464ded1a34f7fa046ce7dfdb7e7f57a

**Review by ValuedMammal — APPROVED** [review]
ACK a53f07be1464ded1a34f7fa046ce7dfdb7e7f57a may need a rebase to kick CI.

In a53f07b: 

> test: add tests for  trait of operations on various data structures, including conflict scenarios and checking for empty structures

It would be better if the commit message stated that we're testing `Merge`.

**Comment by notmandatory** [comment]
Please rebase to pick up CI changes from #1803.

**Comment by notmandatory** [comment]
I did the rebase. :-)
